tests/integration/tables/test_tpp.py

```diff
@@ -220,3 +223,77 @@ def test_addresses_imd_quintile(in_memory_engine):
         },
         {"patient_id": 6, "imd_quintile": "unknown", "imd_decile": "unknown"},
     ]
+
+
+def test_apcs_all_diagnoses_cant_match_string():
+    with pytest.raises(TypeError):
+        tpp.apcs.all_diagnoses == "I000"
+    with pytest.raises(TypeError):
+        tpp.apcs.all_diagnoses != "I000"
+
+
+def test_apcs_all_diagnoses_matches_icd10_prefix_or_icd10_code(in_memory_engine):
```

Comment:
I don't think this is quite the right place for these tests (which obviously you couldn't be expected to know!). This file is for testing the logic of helper methods which are specific to tables in the `tpp` schema.

I think the right place for exercising this logic is in what we call the "spec" tests, which you probably haven't seen yet. They live in `tests/spec` and also get built as [part of the docs](https://docs.opensafely.org/ehrql/reference/features/). We had grand plans for how great automatically creating a readable specification from test cases was going to be but we never fully [followed through](https://github.com/opensafely-core/ehrql/issues/1402).

One of the nice things about the spec tests is that they're automatically executed over all the query engines so we can enforce consistent behaviour there.

You'll need someone to talk you through these (and I'll need to do some serious memory-refreshing beforehand!) but the basic idea would be to add a column of the new type to the [test schema](https://github.com/opensafely-core/ehrql/blob/562d5baf67107d5f4ef8156c58d6ffc954f5342a/tests/spec/tables.py). And then add a new test along the lines of [`code_series_ops.test_containment`](https://github.com/opensafely-core/ehrql/blob/562d5baf67107d5f4ef8156c58d6ffc954f5342a/tests/spec/code_series_ops/test_containment.py) which exercises the new type.

We _could_ make the argument that all we really need to test here is that the ehrQL constructs compile to the correct query model structures as we're not actually adding any new operations to the query model, and the existing query model operations are fully tested elsewhere. That could be done entirely in the unit tests without hitting the database.

But there's something more reassuring about hitting a real database and seeing it give the correct results, so I think the spec tests are probably the way to go.